tests/frame_manager_test.go

```diff
@@ -41,32 +41,18 @@ func TestWaitForFrameNavigationWithinDocument(t *testing.T) {
 
 					resp := p.Goto(tb.staticURL("/nav_in_doc.html"), tb.toGojaValue(&common.FrameGotoOptions{
 						WaitUntil: common.LifecycleEventNetworkIdle,
-						Timeout:   timeout, // interpreted as ms
+						Timeout:   time.Duration(timeout.Milliseconds()), // interpreted as ms
 					}))
 					require.NotNil(t, resp)
-
-					// Callbacks that are initiated internally by click and WaitForNavigation
-					// need to be called from the event loop itself, otherwise the callback
-					// doesn't work. The await below needs to first return before the callback
-					// will resolve/reject.
-					var wfnPromise, cPromise *goja.Promise
-					err := tb.await(func() error {
-						wfnPromise = p.WaitForNavigation(tb.toGojaValue(&common.FrameWaitForNavigationOptions{
-							Timeout: timeout, // interpreted as ms
-						}))
-						cPromise = p.Click(tc.selector, nil)
-
-						assert.Equal(t, goja.PromiseStatePending, wfnPromise.State())
-						assert.Equal(t, goja.PromiseStatePending, cPromise.State())
-
-						return nil
-					})
```

Comment:
I wonder what is the difference between:
* Initiating two promises in the `await` closure, vs,
* Putting them in a `promiseAll`?

Is it about making it as much closer to the Javascript implementation as possible?

Reply:
> Initiating two promises in the await closure 

`await` is misleadingly named (and I will likely make a PR once I figure a better name). `await` here is literally "start the event loop" which will wait until it doesn't have any more work or error occured. 

Calling `await` *within* `await` starts the event loop within the event loop, which is not supported - and doesn't even have a defined behaviour. It worked in this particular case, because the outer await is basically doing nothing, but in general *should* not be done.

Given that I am actually moving `goto` to be async (already done, just figuring out how to make the test less ... painful to write) I am in practice building up the codebase to make writing tests with async parts easier. 

> much closer to the Javascript implementation as possible

This is another argument and arguably a lot of the tests should just run `js` code instead. But I am not going to rewrite that ;). Also at least some need (or at least use) access to things that aren't available from JavaScript and should not be available. So from that perspective things such as PromiseThen and promiseAll will be needed (or will need to be written on case by case basis ;).

Reply:
Thanks for explaining 👍